### DISCUSSION TITLE
How can I use `manif` with `Ceres`?

### DISCUSSION THREAD

**Discussion by fghoussen** [artivis/manif#315] [Q&A]
How can I use `manif` with `Ceres`?

This question is in connection with #314, and this specific quote from it `Imagine you are in SE3. You have a pose (R,T) that you want to update with a displacement, obtained by integrating angular and linear velocities (w,v) over a period dt.`. Here I do not have a `(w, v)` but I need to estimate a jacobian.

`Ceres` is a software package designed to minimize cost functions: basically, you pass a definition of `f` and `J=df/dx` to `Ceres`, and, `Ceres` does the solve for you. `Ceres` looks to stand in the Euclidean world and have no notion of the Lie theory. Note that in practice, in `Ceres`, you _just_ estimate `f` and `J` from a given `x`, and then, pass the result back to `Ceres` (you do _not_ multiply `J` by `dx` yourself to follow the Gradient Descent: AFAIU, `Ceres` does it for you).

When `f` is defined on SE3, starting from a constraint based on a pose `(R, T)` for example, it's not clear to me how I can use `manif` with `Ceres`? Can the answer be possibly: no, you can't?

There are 2 things I stumble on:
- `J` must be evaluated from a given `x`: I guess I can do that estimation with `manif` even if it's not yet clear to me how. Could somebody illustrate how?
- For me, the multiply `.` in `J.dx` is where the exponential-map magic occurs in backend: this seems to be done by `Ceres` without any way to override it... So I would say I can not benefit from Lie theory using `Ceres`: is this correct?

If somebody could shed some light (even partailly), that would be helpful!...

Is there an example of this I could start from?

Note: as lots of things are still blurry, it's difficult to formulate precise questions... So I hope my questions are precise enough...

**Comment by joansola** [comment]
Look here https://artivis.github.io/manif/cpp/pages.html

**Reply by fghoussen** [reply]
Thanks! Is there an example of this?

**Comment by joansola** [comment]
See this for the theory:
[jacobians.pdf](https://github.com/user-attachments/files/17519025/jacobians.pdf)

and see the way we interface with Ceres in our library WOLF, which does not use manif.

Wolf is here: 
http://mobile_robotics.pages.iri.upc-csic.es/wolf_projects/wolf_lib/wolf-doc-sphinx/

You can look at manif's tests with ceres, here

https://github.com/artivis/manif/tree/devel/test/ceres

You will need to dig into the theory and the code, I think we gave you all the insights that you need.

If you want to use manif with ceres, you will have to come up with appropriate hacks so that you can use manif Jacobians in ceres. These hacks usually consist on defining the Jacobian of the local parametrization (in ceres' class Manifold) to be the identity, but padded with zeros to match the different matrix sizes. To understand this very last comment, you need to dig into the theory and code, as I said.

**Reply by fghoussen** [reply]
Thanks!

**Reply by joansola** [reply]
implementation-wise, @artivis knows better.

**Reply by artivis** [reply]
Have a look at e.g. [gtest_se2_ceres_autodiff.cpp](https://github.com/artivis/manif/blob/devel/test/ceres/gtest_se2_ceres_autodiff.cpp) which implements a couple small examples.

manif is Ceres compatible in that it support Ceres' autodiff (Jet type) as stated in the [README](https://github.com/artivis/manif/tree/devel?tab=readme-ov-file#note).
